### PR TITLE
VP-6350: Add protection against category hierarchy cycles

### PR DESCRIPTION
**Pull request by yecli**
Moving categories to its children is prohibited now.

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Data/Validation/ListEntriesMoveRequestValidator.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                if (targetCategory == null)
+                {
+                    context.AddFailure($"Destination category does not exist.");
+                    return;
+                }
+
+                foreach (var movedCategory in moveInfo.ListEntries.Where(x => x.Type.EqualsInvariant(CategoryListEntry.TypeName)))
+                {
+                    var movedCategoryPath = string.Join("/", movedCategory.Outline);
+                    var targetCategoryPath = string.Join("/", targetCategory.Outlines.FirstOrDefault());
```
Comment (tatarincev) [inline src/VirtoCommerce.CatalogModule.Data/Validation/ListEntriesMoveRequestValidator.cs:36]: Why you do not use the `targetCategory.Outline ` here?

File: src/VirtoCommerce.CatalogModule.Data/Services/CategoryMover.cs [hunk 2]
```diff
@@ -51,5 +56,21 @@ public override async Task<List<Category>> PrepareMoveAsync(ListEntriesMoveReque
...
+
+        protected virtual async Task ValidateOperationArguments(ListEntriesMoveRequest moveInfo)
+        {
+            ValidateParameters(moveInfo);
+
+            var validator = new ListEntriesMoveRequestValidator(_categoryService);
+            await validator.ValidateAndThrowAsync(moveInfo);
+        }
+
+        private void ValidateParameters(ListEntriesMoveRequest moveInfo)
```
Comment (tatarincev) [inline src/VirtoCommerce.CatalogModule.Data/Services/CategoryMover.cs:68]: I think is better to disable this stupid validation rule

File: src/VirtoCommerce.CatalogModule.Data/Validation/ListEntriesMoveRequestValidator.cs (hunk 1 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Data/Validation/ListEntriesMoveRequestValidator.cs]: Done, though it increases the complexity of mocking, as this property is calculated with recursion.

File: src/VirtoCommerce.CatalogModule.Data/Services/CategoryMover.cs (hunk 2 above)
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Data/Services/CategoryMover.cs]: Done

**Review by tatarincev — APPROVED** [review]

**Pull request merged**